src/singletons/Crashpad.cpp

```diff
@@ -0,0 +1,216 @@
+/// to be able to read them without loading the settings.
+///
+/// The flags are saved in the `RECOVERY_FILE` in their binary representation.
+std::optional<CrashRecovery::Flags> readFlags(const Paths &paths)
+{
+    using Flag = CrashRecovery::Flag;
+
+    QFile file(QDir(paths.crashdumpDirectory).filePath(RECOVERY_FILE));
+    if (!file.open(QFile::ReadOnly))
+    {
+        return std::nullopt;
+    }
+
+    auto line = file.readLine(64);
+
+    static_assert(std::is_same_v<std::underlying_type_t<Flag>, qulonglong>);
+
+    bool ok = false;
+    auto value = static_cast<Flag>(line.toULongLong(&ok));
+    if (!ok)
+    {
+        return std::nullopt;
+    }
+
+    return value;
+}
+
+bool canRestart(const Paths &paths)
+{
+#ifdef NDEBUG
+    const auto &args = chatterino::getArgs();
+    bool noBadArgs =
+        !args.isFramelessEmbed && !args.shouldRunBrowserExtensionHost;
+
+    return noBadArgs && readFlags(paths)
+                            .value_or(CrashRecovery::Flag::None)
+                            .has(CrashRecovery::Flag::DoCrashRecovery);
+#else
+    (void)paths;
+    return false;
+#endif
+}
+
+std::string encodeArguments()
+{
+    std::string args;
+    for (auto arg : getArgs().currentArguments())
+    {
+        if (!args.empty())
+        {
+            args.push_back('+');
+        }
+        args += arg.replace(u'+', u"++"_s).toStdString();
+    }
+    return args;
+}
+
+}  // namespace
+
+namespace chatterino {
+
+using namespace std::string_literals;
+
+void CrashRecovery::initialize(Settings & /*settings*/, Paths &paths)
+{
+    auto flags = readFlags(paths);
+    if (flags)
+    {
+        this->currentFlags_ = *flags;
+    }
+    else
+    {
+        // By default, we don't restart after a crash.
+        this->updateFlags(Flag::None);
+    }
+}
+
+CrashRecovery::Flags CrashRecovery::recoveryFlags() const
+{
+    return this->currentFlags_;
+}
+
+void CrashRecovery::updateFlags(CrashRecovery::Flags flags)
+{
+    this->currentFlags_ = flags;
+
+    QFile file(QDir(getPaths()->crashdumpDirectory).filePath(RECOVERY_FILE));
+    if (!file.open(QFile::WriteOnly | QFile::Truncate))
+    {
+        qCWarning(chatterinoApp) << "Failed to open" << file.fileName();
+        return;
+    }
+
+    static_assert(std::is_same_v<std::underlying_type_t<Flag>, qulonglong>);
+    file.write(QByteArray::number(
+        static_cast<qulonglong>(this->currentFlags_.value())));
+}
+
+#ifdef CHATTERINO_WITH_CRASHPAD
+std::unique_ptr<crashpad::CrashpadClient> installCrashHandler()
+{
+    // Currently, the following directory layout is assumed:
+    // [applicationDirPath]
+    //  ├─chatterino(.exe)
+    //  ╰─[crashpad]
+    //     ╰─crashpad_handler(.exe)
+    // TODO: The location of the binary might vary across platforms
+    auto crashpadBinDir = QDir(QApplication::applicationDirPath());
+
+    if (!crashpadBinDir.cd("crashpad"))
+    {
+        qCDebug(chatterinoApp) << "Cannot find crashpad directory";
```

Comment:
Let's make a new logging category called `crashhandler`
Rename this file to CrashHandler